src/client/page-navigation-watch.js

```diff
@@ -0,0 +1,138 @@
+import { isShadowUIElement, isAnchorElement, isFormElement, closest } from './utils/dom';
+import ElementSandbox from './sandbox/node/element';
+import * as windowsStorage from './sandbox/windows-storage';
+import domProcessor from './dom-processor/index';
+import nextTick from './utils/next-tick';
+import nativeMethods from './sandbox/native-methods';
+
+
+export default class PageNavigationWatch extends EventEmiter {
+    constructor (eventSandbox, codeInstrumentation, elementSandbox) {
+        super();
+
+        this.PAGE_NAVIGATION_TRIGGERED_EVENT = 'hammerhead|event|page-navigation-triggered';
+
+        this.lastLocationValue = window.location.toString();
+
+        this._locationWatch(codeInstrumentation);
+        this._linkWatch(eventSandbox);
+        this._formWatch(elementSandbox, eventSandbox);
+    }
+
+    _formWatch (elementSandbox, eventSandbox) {
+        var onFormSubmit = form => {
+            var targetWindow = PageNavigationWatch._getTargetWindow(form);
+
+            PageNavigationWatch._onNavigationTriggeredInWindow(targetWindow, form.action);
+        };
+
+        // NOTE: this raises when form.submit() was called
+        elementSandbox.on(elementSandbox.BEFORE_FORM_SUBMIT, e => onFormSubmit(e.form));
+
+        // NOTE: this raises when form submitting was triggered by a submit button click
+        eventSandbox.listeners.initElementListening(window, ['submit']);
+        eventSandbox.listeners.addInternalEventListener(window, ['submit'], e => {
+            var prevented = false;
+
+            if (!isFormElement(e.target))
+                return;
+
+            var onPreventDefault = preventedEvent => {
+                prevented = prevented || preventedEvent === e;
+            };
+
+            eventSandbox.on(eventSandbox.EVENT_PREVENTED_EVENT, onPreventDefault);
+
+            nextTick()
+                .then(() => {
+                    eventSandbox.off(eventSandbox.EVENT_PREVENTED_EVENT, onPreventDefault);
+
+                    // NOTE: the event keep the defaultPrevented flag between event handlers in all
+                    // browsers except IE. In IE it is reset to false before the next handler execution.
+                    if (!e.defaultPrevented && !prevented)
+                        onFormSubmit(e.target);
+                });
+        });
+    }
+
+    static _getTargetWindow (el) {
+        var targetWindow = window;
+        var target       = nativeMethods.getAttribute.call(el, domProcessor.getStoredAttrName('target')) ||
+                           nativeMethods.getAttribute.call(el, 'target');
+
+        if (target) {
+            if (!ElementSandbox._isKeywordTarget(target))
+                targetWindow = windowsStorage.findByName(target) || window;
+            else if (target === '_top' || target === '_blank')
+                targetWindow = window.top;
+            else if (target === '_parent')
+                targetWindow = window.parent;
```

Comment:
If a window does not have a parent, its parent property is a reference to itself (
[proof](https://developer.mozilla.org/en/docs/Web/API/Window/parent)). Do you know a case when `window.parent === null` ?